### PR TITLE
Multiblocks are coming back! Bleeding edge

### PR DESCRIPTION
**Pull request by liach**
Fixes a few bugs: #1353 #1339 #1356 #1347 #1338

### CODE REVIEW

**Comment by liach** [comment]
Fixes #430, too

File: src/main/java/mods/railcraft/common/blocks/machine/MachineTileRegistry.java [hunk 1]
```diff
@@ -41,15 +38,15 @@
...
+        RailcraftRegistry.register(TileBlastFurnace.class, "blast_furnace");
+        RailcraftRegistry.register(TileCokeOven.class, "coke_oven");
+        RailcraftRegistry.register(TileRockCrusher.class, "rock_crusher");
+        RailcraftRegistry.register(TileTankWater.class, "water_tank");
+        RailcraftRegistry.register(TileSteamOven.class, "steam_oven");
+        RailcraftRegistry.register(TileSteamTrapManual.class, "steam_trap_manual");
+        RailcraftRegistry.register(TileSteamTrapAuto.class, "steam_trap_auto");
 
         // Beta
         GameRegistry.registerTileEntity(TileEngineSteamHobby.class, "RCEngineSteamHobby");
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/blocks/machine/MachineTileRegistry.java]: Shouldn't these be updated too?

File: src/main/java/mods/railcraft/common/blocks/multi/TileBlastFurnace.java [hunk 2]
```diff
@@ -149,28 +149,22 @@ public static void placeBlastFurnace(World world, BlockPos pos, ItemStack input,
...
                 break;
             case 'B':
             case 'W':
-                if (block == RailcraftBlocks.MACHINE_ALPHA.block() && meta == getBlockMetadata())
+                if (self == state)
                     return true;
                 break;
             case 'A':
-                if (block.isAir(state, worldObj, pos) || state.getMaterial() == Material.LAVA)
+                if (world.isAirBlock(pos) || state.getMaterial() == Material.LAVA)
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/blocks/multi/TileBlastFurnace.java]: Its more efficient to ask the block directly if it is air, rather than the world.

File: src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java [hunk 3]
```diff
@@ -174,8 +170,8 @@ public int getLightValue() {
     }
 
     private void processActions() {
-        paused = actions.stream().anyMatch(a -> a == Actions.PAUSE);
-        actions.clear();
+//        paused = actions.stream().anyMatch(a -> a == Actions.PAUSE);
+//        actions.clear();
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java]: Why is the action code commented out?

File: src/main/java/mods/railcraft/common/blocks/single/TileTradeStation.java [hunk 4]
```diff
@@ -125,10 +122,10 @@ private boolean attemptTrade(List<EntityVillager> villagers, int tradeSet) {
                 if (recipe.isRecipeDisabled())
                     continue;
                 //noinspection ConstantConditions
-                if (recipe.getItemToBuy() != null && !InvTools.isItemLessThanOrEqualTo(recipe.getItemToBuy(), buy1))
+                if (recipe.getItemToBuy() != InvTools.emptyStack() && !InvTools.isItemLessThanOrEqualTo(recipe.getItemToBuy(), buy1))
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/blocks/single/TileTradeStation.java]: This probably isn't 100% safe, use isEmpty(stack) in all these tests instead.

File: src/main/java/mods/railcraft/common/carts/MinecartHooks.java [hunk 5]
```diff
@@ -206,18 +206,17 @@ public void onEntityCollision(EntityMinecart cart, Entity other) {
...
     }
 
     @Override
     public AxisAlignedBB getMinecartCollisionBox(EntityMinecart cart) {
-        double yaw = Math.toRadians(cart.rotationYaw);
-        double diff = ((CART_LENGTH - CART_WIDTH) / 2.0) + MinecartHooks.COLLISION_EXPANSION;
-        double x = diff * Math.abs(Math.cos(yaw));
-        double z = diff * Math.abs(Math.sin(yaw));
-        return cart.getEntityBoundingBox().expand(x, MinecartHooks.COLLISION_EXPANSION, z);
+        return cart.getEntityBoundingBox().expand(MinecartHooks.COLLISION_EXPANSION, 0, MinecartHooks.COLLISION_EXPANSION);
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/carts/MinecartHooks.java]: Does this break the Tunnel Bore collision? Hmm....not sure why I do this.

File: src/main/java/mods/railcraft/common/carts/MinecartHooks.java [hunk 6]
```diff
@@ -206,18 +206,17 @@ public void onEntityCollision(EntityMinecart cart, Entity other) {
     public AxisAlignedBB getCollisionBox(EntityMinecart cart, Entity other) {
         if (other instanceof EntityItem && RailcraftConfig.doCartsCollideWithItems())
             return other.getEntityBoundingBox();
-        if (other instanceof EntityPlayer)
-            return other.canBePushed() ? other.getEntityBoundingBox() : null; //            return other.boundingBox.contract(COLLISION_EXPANSION, 0, COLLISION_EXPANSION);
-        return null;
+        return other.canBePushed() ? other.getEntityBoundingBox() : null; //            return other.boundingBox.contract(COLLISION_EXPANSION, 0, COLLISION_EXPANSION);
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/carts/MinecartHooks.java:209]: This allows any entity to push minecarts yes? Probably not desirable.

File: src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java [hunk 7]
```diff
@@ -18,11 +18,6 @@ public SlotPassThrough(IInventory inv, int slotIndex, int posX, int posY) {
         super(inv, slotIndex, posX, posY);
     }
 
-    @Override
-    public int getSlotStackLimit() {
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java:22]: Why is this removed? This specific Slot should mirror the underlying inventory.

File: src/main/java/mods/railcraft/common/items/RailcraftItems.java [hunk 8]
```diff
@@ -168,13 +168,13 @@
     }},
     TRACK_PARTS(ItemTrackParts::new, "track_parts", () -> "ingotIron"),
     TURBINE_BLADE(ItemTurbineBlade::new, "turbine_blade", () -> "ingotSteel") {{
-        conditions().add(EnumMachineAlpha.TURBINE);
+//        conditions().add(EnumMachineAlpha.TURBINE);
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/items/RailcraftItems.java]: These conditions need updating.

File: src/main/java/mods/railcraft/common/modules/ModuleFactory.java [hunk 9]
```diff
@@ -366,7 +351,7 @@ private void registerCrushedOreRecipe(ItemStack ore, ItemStack dust) {
             public void postInit() {
                 if (!EquipmentVariant.ROLLING_MACHINE_POWERED.isAvailable())
                     RollingMachineCraftingManager.copyRecipesToWorkbench();
-                if (!EnumMachineAlpha.BLAST_FURNACE.isAvailable() || RailcraftConfig.forceEnableSteelRecipe())
+                if (/*!EnumMachineAlpha.BLAST_FURNACE.isAvailable() || */RailcraftConfig.forceEnableSteelRecipe())
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/modules/ModuleFactory.java]: Update?

File: src/main/java/mods/railcraft/common/modules/ModuleResources.java [hunk 10]
```diff
@@ -75,8 +75,8 @@ public void preInit() {
                     if (RailcraftConfig.isSubBlockEnabled(type.getTag())) {
                         initMetalBlock(Metal.STEEL);
 
-                        if (EnumMachineAlpha.BLAST_FURNACE.isAvailable())
-                            RailcraftCraftingManager.blastFurnace.addRecipe(new ItemStack(Blocks.IRON_BLOCK), false, false, 11520, EnumGeneric.BLOCK_STEEL.getStack());
+//                        if (EnumMachineAlpha.BLAST_FURNACE.isAvailable())
+//                            RailcraftCraftingManager.blastFurnace.addRecipe(new ItemStack(Blocks.IRON_BLOCK), false, false, 11520, EnumGeneric.BLOCK_STEEL.getStack());
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/modules/ModuleResources.java]: I assume this moved somewhere else.

File: src/main/java/mods/railcraft/common/plugins/forge/WorldPlugin.java [hunk 11]
```diff
@@ -48,22 +46,34 @@ public static Block getBlock(IBlockAccess world, BlockPos pos) {
 
     @Nullable
     public static TileEntity getBlockTile(IBlockAccess world, BlockPos pos) {
-        if (pos.getY() < 0)
-            return null;
-        return world.getTileEntity(pos);
+        // see flowerpot source code
+        return world instanceof ChunkCache ? ((ChunkCache) world).getTileEntity(pos, Chunk.EnumCreateEntityType.CHECK) : world.getTileEntity(pos);
```
Comment (CovertJaguar) [inline src/main/java/mods/railcraft/common/plugins/forge/WorldPlugin.java]: Shouldn't we still leave the world bounds check? The game used to crash if you passed y == -1.

**Review by CovertJaguar — COMMENTED** [review]
Generally looks good, just some really minor issues.

Great job!

File: src/main/java/mods/railcraft/common/blocks/single/TileTradeStation.java (hunk 4 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/blocks/single/TileTradeStation.java]: Yeah made that fault.

File: src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java (hunk 3 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java]: Because I think they may be useless in the immediate future

File: src/main/java/mods/railcraft/common/carts/MinecartHooks.java (hunk 5 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/carts/MinecartHooks.java]: I think I added it when I was trying to fix the locomotive hitting people thing.

File: src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java (hunk 7 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java:22]: That is set to a field in the basic slot class.

File: src/main/java/mods/railcraft/common/items/RailcraftItems.java (hunk 8 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/items/RailcraftItems.java]: yep.

File: src/main/java/mods/railcraft/common/modules/ModuleFactory.java (hunk 9 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/modules/ModuleFactory.java]: Move it into blast furnace class?

File: src/main/java/mods/railcraft/common/plugins/forge/WorldPlugin.java (hunk 11 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/plugins/forge/WorldPlugin.java]: I will try to make this into an error message on log instead.

File: src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java (hunk 7 above)
Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java:22]: Yes, but this specific slot class is different. It is designed to simply look at the underlying inventory and get the info from there.

File: src/main/java/mods/railcraft/common/modules/ModuleFactory.java (hunk 9 above)
Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/modules/ModuleFactory.java]: Sure

File: src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java (hunk 3 above)
Comment (CovertJaguar) [reply src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java]: K

File: src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java (hunk 7 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java:22]: Hmm. Involves some design thing. Will look into it more.

File: src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java (hunk 3 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/blocks/multi/TileMultiBlockOven.java]: Will restore it.

**Comment by liach** [comment]
@CovertJaguar For your information, I would not be able to do much coding till March 24th due to life issues. Sorry for that.

**Comment by border999** [comment]
@liach Welcome back

File: src/main/java/mods/railcraft/common/carts/MinecartHooks.java (hunk 6 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/carts/MinecartHooks.java:209]: Hmm, I think I added that in order to fix the locomotive hitting thing.

File: src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java (hunk 7 above)
Comment (liach) [reply src/main/java/mods/railcraft/common/gui/slots/SlotPassThrough.java:22]: After investigation, the current code preserves the original behavior, that since a hard limit is not present the original `inventory.getInventoryStackLimit()` is always returned.

```java
    @Override
    public final int getSlotStackLimit() {
        int max = super.getSlotStackLimit();
        return stackLimit < 0 ? max : Math.min(max, stackLimit); // issue #1347
    }
```
Where `stackLimit` is defaulted to `-1`.

**Review by CovertJaguar — APPROVED** [review]
